clang/runtime/dpct-rt/include/debug/schema.hpp

```diff
@@ -0,0 +1,391 @@
+  ValType get_val_type() { return ValTy; }
+  size_t get_val_size() { return ValSize; }
+  size_t get_type_size() { return TypeSize; }
+  bool is_basic_type() { return IsBasicType; }
+  const std::string &get_type_name() { return TypeName; }
+  bool is_virtual_type() { return IsVirtual; }
+  size_t get_feild_num() { return FieldNum; }
+  std::vector<std::shared_ptr<Schema>> &get_type_member() { return Members; }
+  size_t get_offset() { return Offset; }
+  const std::string &get_var_name() { return VarName; }
+
+private:
+  // Common Part
+  std::string TypeName = "";                // namespace + class type + name;
+  schema_type SchemaTy = schema_type::TYPE; // Data or Type
+  std::string VarName = "";
+  bool IsBasicType = true;
+  std::string FilePath;
+  /* Class Type only */
+  bool IsVirtual = false;
+  size_t FieldNum = 0;
+  size_t TypeSize = 0;
+  std::vector<std::shared_ptr<Schema>> Members;
+  /* Data member only */
+  ValType ValTy = ValType::SCALAR;
+  size_t Offset = 0;
+  size_t ValSize = 0;
+  MemLoc Location = MemLoc::NONE;
+};
+
+// data, test_namespace::A + varName,
+inline std::pair<std::string, std::shared_ptr<Schema>>
+create_schema_struct(const std::string &TypeName, schema_type SchemaTy,
+                     size_t FieldNum, bool IsVirtual, size_t TypeSize,
+                     const std::string &FilePath) {
+  return std::pair<std::string, std::shared_ptr<Schema>>(
+      TypeName, std::make_shared<Schema>(TypeName, SchemaTy, FieldNum,
+                                         IsVirtual, TypeSize, FilePath));
+}
+
+inline void create_schema_member(std::shared_ptr<Schema> TypeSchemaStruct,
+                                 const std::string &VarName,
+                                 const std::string &TypeName, size_t type_size,
+                                 bool IsBasicType, ValType ValTy,
+                                 size_t ValSize, size_t Offset,
+                                 MemLoc Location) {
+  std::shared_ptr<Schema> DataMember =
+      std::make_shared<Schema>(VarName, TypeName, type_size, IsBasicType, ValTy,
+                               ValSize, Offset, Location);
+  TypeSchemaStruct->add_data_schema(DataMember);
+}
+
+inline std::pair<std::string, std::shared_ptr<Schema>>
+create_schema_var(const std::string &VarName, const std::string &TypeName,
+                  size_t type_size, schema_type SchemaTy, bool IsBasicType,
+                  ValType ValTy, size_t ValSize, MemLoc Location) {
+  return std::pair<std::string, std::shared_ptr<Schema>>(
+      TypeName,
+      std::make_shared<Schema>(VarName, TypeName, type_size, SchemaTy,
+                               IsBasicType, ValTy, ValSize, Location));
+}
+
+inline std::shared_ptr<Schema> get_type_schema(dpct_json::value &v) {
+  auto obj = v.get_value<dpct_json::object>();
+  std::string schema_name = obj.get("TypeName").get_value<std::string>();
+
+  std::string schema_type_name =
+      obj.get("schema_type").get_value<std::string>();
```

Comment:
obj.get("schema_type")， I think we need unify the style of schema keyword.
"TypeName" "FieldNum" "IsVirtual", most keyword is named as CamelCase style, only this one is slash.
I prefer "SchemaType"

Reply:
Updated